From 3a6abb338c3e29be8f48ef231678336d8e6de13f Mon Sep 17 00:00:00 2001
From: Pankaj Dev <pankaj.dev@st.com>
Date: Wed, 8 Dec 2021 18:25:11 +0530
Subject: [PATCH 51/65] usb: dwc3: Add support for STM32 platform in
 dwc3-generic

Add support for STM32 family platforms, currently for STM32MP2

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 8a3aa6498758e7bc17e0ccfe3fb191a19e5fb255 usb: dwc3: Add support for STM32 platform in dwc3-generic]
Change-Id: I94a8da41549bb4cc0ffb892d1a7712e5d8ed5ccc
Signed-off-by: Pankaj Dev <pankaj.dev@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/303888
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/usb/dwc3/dwc3-generic.c | 68 ++++++++++++++++++++++++++++++++-
 1 file changed, 66 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-generic.c b/drivers/usb/dwc3/dwc3-generic.c
index c09014aec60..13d1d8ca7e9 100644
--- a/drivers/usb/dwc3/dwc3-generic.c
+++ b/drivers/usb/dwc3/dwc3-generic.c
@@ -7,15 +7,22 @@
  * Based on dwc3-omap.c.
  */
 
+#define LOG_CATEGORY UCLASS_USB
+
 #include <dm.h>
 #include <reset.h>
 #include <asm/gpio.h>
+#include <dm/device_compat.h>
 #include <dm/lists.h>
+#include <linux/bitfield.h>
 #include <linux/delay.h>
 #include <linux/usb/gadget.h>
 #include <power/regulator.h>
+#include <regmap.h>
+#include <syscon.h>
 #include <usb/xhci.h>
 #include "core.h"
+
 #include "dwc3-generic.h"
 #include "gadget.h"
 
@@ -163,7 +170,7 @@ static int dwc3_generic_of_to_plat(struct udevice *dev)
 
 	plat->maximum_speed = usb_get_maximum_speed(node);
 	if (plat->maximum_speed == USB_SPEED_UNKNOWN) {
-		pr_info("No USB maximum speed specified. Using super speed\n");
+		dev_info(dev, "No USB maximum speed specified. Using super speed\n");
 		plat->maximum_speed = USB_SPEED_SUPER;
 	}
 
@@ -173,7 +180,7 @@ static int dwc3_generic_of_to_plat(struct udevice *dev)
 		node = dev_ofnode(dev->parent);
 		plat->dr_mode = usb_get_dr_mode(node);
 		if (plat->dr_mode == USB_DR_MODE_UNKNOWN) {
-			pr_err("Invalid usb mode setup\n");
+			dev_err(dev, "Invalid usb mode setup\n");
 			return -ENODEV;
 		}
 	}
@@ -488,6 +495,62 @@ struct dwc3_glue_ops qcom_ops = {
 	.glue_configure = dwc3_qcom_glue_configure,
 };
 
+void dwc3_stm32_glue_configure(struct udevice *dev, int index, enum usb_dr_mode mode)
+{
+#define SYSCFG_USB3DRCR_USB2ONLYH_MASK        BIT(3U) /*!< 0x00000008 : USB2-only Mode for Host */
+#define SYSCFG_USB3DRCR_USB2ONLYD_MASK        BIT(4U) /*!< 0x00000010 : USB2-only Mode for Device */
+	int ret;
+	struct regmap *regmap;
+	struct dwc3_glue_data *glue = dev_get_plat(dev);
+	u32 syscfg_usb3drcr_reg_off;
+	bool usb2only_conf;
+	regmap = syscon_regmap_lookup_by_phandle(dev, "st,syscfg");
+	if (IS_ERR(regmap)) {
+		dev_err(dev, "unable to find regmap\n");
+		return;
+	}
+	ret = ofnode_count_phandle_with_args(ofnode_first_subnode(dev_ofnode(dev)),
+					     "phys", "#phy-cells", 0);
+	if (ret < 1) {
+		dev_err(dev, "unable to find phys\n");
+		return;
+	}
+	if (ret < 2)
+		usb2only_conf = true;
+	else
+		usb2only_conf = false;
+	dev_info(dev, "configured in %s mode\n", usb2only_conf ? "usb2" : "usb3");
+	ret = dev_read_u32_index(dev, "st,syscfg", 1, &syscfg_usb3drcr_reg_off);
+	if (ret) {
+		dev_err(dev, "Can't get sysconfig usb3drcr offset (%d)\n", ret);
+		return;
+	}
+	dev_dbg(dev, "syscfg-usb3drcr-reg offset 0x%x\n", syscfg_usb3drcr_reg_off);
+	ret = regmap_update_bits(regmap, syscfg_usb3drcr_reg_off, SYSCFG_USB3DRCR_USB2ONLYD_MASK |
+			SYSCFG_USB3DRCR_USB2ONLYH_MASK,
+			FIELD_PREP(SYSCFG_USB3DRCR_USB2ONLYD_MASK, usb2only_conf ? 1 : 0) |
+			FIELD_PREP(SYSCFG_USB3DRCR_USB2ONLYH_MASK, usb2only_conf ? 1 : 0));
+	if (ret) {
+		dev_err(dev, "regmap_write error: %d\n", ret);
+		return;
+	}
+	// Assert + Deassert Reset for DWC3-ctrl to sample syscfg settings
+	ret = reset_assert_bulk(&glue->resets);
+	if (ret) {
+		dev_err(dev, "reset_assert_bulk error: %d\n", ret);
+		return;
+	}
+	ret = reset_deassert_bulk(&glue->resets);
+	if (ret) {
+		dev_err(dev, "reset_deassert_bulk error: %d\n", ret);
+		return;
+	}
+}
+
+struct dwc3_glue_ops stm32_ops = {
+	.glue_configure = dwc3_stm32_glue_configure,
+};
+
 static int dwc3_rk_glue_get_ctrl_dev(struct udevice *dev, ofnode *node)
 {
 	*node = dev_ofnode(dev);
@@ -707,6 +770,7 @@ static const struct udevice_id dwc3_glue_ids[] = {
 	{ .compatible = "fsl,imx8mq-dwc3" },
 	{ .compatible = "intel,tangier-dwc3" },
 	{ .compatible = "samsung,exynos850-dwusb3" },
+	{ .compatible = "st,stm32mp25-dwc3", .data = (ulong)&stm32_ops },
 	{ }
 };
 
-- 
2.51.0

